File: python/Neetcode/SlidingWindow/Medium/3.LongestSubstring.py
# Given a string s, find the length of the longest substring
# without repeating characters.

def lengthOfLongestSubstring(s):
    window = set()
    left = 0
    result = 0

    for i in range(len(s)):
        while s[i] in window:
            window.remove(s[left])
            left += 1

        window.add(s[i])
        result = max(result, i - left + 1)

    return result

print(lengthOfLongestSubstring("abcabcbb"))
print(lengthOfLongestSubstring("bbbbb"))
print(lengthOfLongestSubstring("pwwkew"))

File: swift/Neetcode/SlidingWindow/Medium/3.LongestSubstring.swift
// Given a string s, find the length of the longest substring
// without repeating characters.

func lengthOfLongestSubstring(_ s: String) -> Int {
    var s = Array(s)
    var window = Set<Character>()
    var result = 0

    for (i, character) in s.enumerated() {
        while window.contains(character) {
            window.remove(s[i - window.count])
        }

        window.insert(character)
        result = max(result, window.count)
    }

    return result
}


print(lengthOfLongestSubstring("abcabcbb"))
print(lengthOfLongestSubstring("bbbbb"))
print(lengthOfLongestSubstring("pwwkew"))